Gutenboarding: Improve a11y and lints

sirreal opened the pull request:
#### Changes proposed in this Pull Request

Add and use Reakit to implement the page selector as a [`Dialog`](https://reakit.io/docs/dialog/). This should improve overall usability of the page selector.

This fixes some lint issues as well as simplifies some styling, for example the full-width page selector overlay. Click outside and <kbd>escape</kbd> close the page selector.


#### Screens

##### Before

![before](https://user-images.githubusercontent.com/841763/71415297-4897aa80-265b-11ea-9176-d7e7d89edb73.gif)

##### After

![after](https://user-images.githubusercontent.com/841763/71415302-4c2b3180-265b-11ea-93de-a3c1a1591ae0.gif)

#### Follow up

- <s>Use [Reakit `Rover`](https://reakit.io/docs/rover/) for tab access to designs and templates in the page selector.</s>
used `Card as="button"` for now.
- Move to css-in-JS which should help clarity and consistency across css/js. There are a lot of duplicated variables and logic right now.
- Transitions are likely to move to React Spring

#### Testing instructions

* Visit [`/gutenboarding`](https://calypso.live/?branch=gutenboarding/improve-a11y-and-lints).
* Complete the first step.
* Navigating between designs should work using keyboard (tabbing).
* On focus / hover over a design there should be a transparent overlay with text.
* Pick a design and verify click outside and <kbd>escape</kbd> work as expected.
* In page layout selector focus should be by default on the first item and tabbing should work.
* Focus should return to the design that was selected.
* Designs are improved (paObgF-RB-p2)

sirreal committed:
Fix classname lint

sirreal committed:
Install reakit

sirreal committed:
Implement with reakit Dialog

sirreal committed:
Fix transitions

sirreal committed:
Fix transition timing regression

sirreal commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -50,7 +50,7 @@ $deisgn-selector-selection-space: 175px;
 	&.exit-done {
 		overflow-y: hidden;
-		height: 100vh;
+		// height: 100vh;
```
```suggestion
```

sirreal committed:
Remove commented line

sirreal commented:
> Sections (~360785 bytes added 📈 [gzipped])

😮 Another wild ICFY report❗️ Why would this affect all those sections so badly?

diegohaz commented on client/landing/gutenboarding/onboarding-block/design-selector/design-card.tsx:
```diff
@@ -20,13 +20,22 @@ interface Props {
 		onClick={ onClick }
 		style={ style }
+		aria-haspopup="dialog"
+		aria-controls={ dialogId }
```
There's no value in adding those aria attributes if the component is not accessible.

> Use Reakit Rover for tab access to designs and templates in the page selector.

I'd not use `Rover` here. `Rover` will apply the roving tabindex method to the cards, which means that, by default, users won't be able to <kbd>Tab</kbd> through the cards. And since they have no clue that they should use arrow keys (as they would with widgets like toolbar, menu etc.), keyboard users may still not be able to navigate through this page.

That could be solved by [overwriting the `tabIndex` prop rendered by `Rover`](https://codesandbox.io/s/reakit-tabbable-rover-ciw69), which would allow users to <kbd>Tab</kbd> and use arrow keys at the same time! 🙌

But that's not the only problem! 😔

`Rover` only adds roving tabindex in one dimension. Sighted users using the keyboard could expect that they could navigate the card list as if it were a grid (bidimensional), since that's what they see. There's an ongoing conversation around adding a `Grid` module to Reakit: https://github.com/reakit/reakit/issues/435

Instead of overcomplicating this, I'd just make this component tabbable.

[`Tabbable`](https://reakit.io/docs/tabbable/), which is used by `Rover` underneath, may help with this, specially because this component is a `div`. `Tabbable` would also make it clickable with <kbd>Enter</kbd> and <kbd>Space</kbd>.

A simpler (and better?) solution though is to just turn this element into a `button`. `<Card as="button"` should do the trick I guess.

Commit:
Enable keyboard accesibility in design and layout selector.
- use HTML button element to render Card
- use span instead of div inside for valid markup

Commit:
Add hover and focus styling for design picker.
- add overlay with text for design selector items
- add a grey border for items of layout selector

simison commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -69,33 +68,42 @@ $deisgn-selector-selection-space: 175px;
 	top: $design-selector-header-vertical-shift + $deisgn-selector-selection-space;
+	width: 100%;
+	background: $white;
+	box-shadow: gray 0 0 12px; // use a sass variable
```
Yeah, indeed something from here;

https://github.com/WordPress/gutenberg/blob/8084e23a8c5364e139790c208a9d9d593b0887b9/packages/base-styles/_colors.scss

Would `$dark-gray-300: #6c7781` work? cc @dwolfe  

Not a blocker for merging, we can update later.

simison replied:
BTW denoting these with `@TODO` is useful as some IDEs highlight them or list them conveniently in one place.

dwolfe replied:
`$dark-gray-300` looks like a good match, let's use that for now.

simison commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -121,6 +129,12 @@ $deisgn-selector-selection-space: 175px;
 
+	&:hover,
+	&:focus {
+		border: 3px solid var( --color-neutral-dark );
```
I'm a little surprised by Calypso color variable here instead of Gutenberg variable. Not a blocker for merging, just something we'll likely need to audit at some point.

razvanpapadopol replied:
It's the color used in design and also in the current page template selector. The closest Gutenberg variable would be `$dark-gray-600: #40464d;` but it has less blue and more grey. 

I agree we need to define together with design team which parts should be using Gutenberg styles and where should come in Calypso theming. 

Also we should take into consideration that mixing color variables will increase the complexity of dark theme support ( https://github.com/Automattic/wp-calypso/issues/38587 )

simison reviewed (approved):
Lookin' good!

Some notes;

- Couldn't get design picker "selected style" (or the text you mention) visible by mouse nor by keyboard by tabbing or with arrow keys. Tabbing does work, just I don't see visually any indication. Tested on Chrome.
- At page selector, could selected page that I tab into turn darker in color, so that tabbing over selected pages wouldn't seem "invisible"? I'd expect same style when I hover already selected page.
   <img width="400" alt="Screenshot 2019-12-30 at 23 54 03" src="https://user-images.githubusercontent.com/87168/71602194-ca0da200-2b5f-11ea-9314-ddc7ba3dfde7.png">
- Scroll position followed nicely tabbed  selection in both design picker and page selector
- Once I tabbed to the last page in the grid, I could see designs peaking: 
    <img width="600" alt="Screenshot 2019-12-30 at 23 56 00" src="https://user-images.githubusercontent.com/87168/71602250-11942e00-2b60-11ea-9c73-33c9017439a8.png">
- Esc cancels all page selections and returns to design picker, just like clicking outside page selector does as well. I expected my page selections to remain tho; they got wiped out.
- I can tab-rotate in pages, but I cannot escape them? Should we tab to "back" button, "domain picker" and other elements after the last page in the grid? Good for follow up PR.

simison commented:
wp-desktop failure seems unrelated — rebasing might help.

dwolfe commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -121,6 +129,12 @@ $deisgn-selector-selection-space: 175px;
+	&:focus {
+		border: 3px solid var( --color-neutral-dark );
+		padding: 0;
+	}
```
We can borrow from the [existing starter page template selector styles](apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/styles/starter-page-templates-editor.scss) - the suggestion below works for page selector cards. Still need to figure out how to work around a specificity collision with `.is-elevated` to use the same style on the design selector cards.

```
	&:focus {
		box-shadow: 0 0 0 1px $white,
					0 0 0 3px $blue-medium-500;
		// Windows High Contrast mode will show this outline, but not the box-shadow.
		outline: 2px solid transparent;
	}

	&:hover {
		border: 3px solid var( --color-neutral-dark );
		padding: 0;
	}
```

razvanpapadopol replied:
Thanks for the suggestion. Opened this issue to try this out: https://github.com/Automattic/wp-calypso/issues/38665

Commit:
Use color variable for box-shadow

Commit:
Fix scss to display design overlay on hover and focus

Commit:
Re-add height limitation for design container when the page layout selector is open.

Commit:
Style selected layout with grey color when focused / hovered so it can be visually identified

razvanpapadopol commented:
> Couldn't get design picker "selected style"

Fixed in 15b1510d7e5f1bd09deaed395bd31887e3937eaa

> tabbing over selected pages wouldn't seem "invisible"? I'd expect same style when I hover already selected page.

Fixed in 9234d31cdad501a08126be6ff1697401ad2d9ca0

>  I could see designs peaking

Fixed in 50a7ff9966b4514e68b79904f96680b61b701351

> I can tab-rotate in pages, but I cannot escape them? Should we tab to "back" button, "domain picker" and other elements after the last page in the grid? Good for follow up PR.

When you are selecting pages (layouts) you're basically in a modal so the focus is captured to that context. So I'd say pressing ESC and then tabbing to another element is somehow expected. But given the fact we don't display an actual overlay on the rest of the page so all the action buttons are visible and clickable, we should discuss this further. Created https://github.com/Automattic/wp-calypso/issues/38664

> Esc cancels all page selections and returns to design picker, just like clicking outside page selector does as well. I expected my page selections to remain tho; they got wiped out.

To be implemented
